File: solutions/cpp/0113.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> ans;
        vector<int> curr;
        dfs(root, sum, curr, ans);
        return ans;
    }

private:
    void dfs(TreeNode* root, int sum, vector<int>& curr, vector<vector<int>>& ans) {
        if (!root) return;
        if (sum == root->val && !root->left && !root->right) {
            curr.push_back(root->val);
            ans.push_back(curr);
            curr.pop_back();
            return;
        }

        curr.push_back(root->val);
        dfs(root->left, sum - root->val, curr, ans);
        dfs(root->right, sum - root->val, curr, ans);
        curr.pop_back();
    }
};

File: solutions/python3/0113.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        ans = []
        self.dfs(root, sum, [], ans)
        return ans

    def dfs(self, root: TreeNode, sum: int, curr: List[int], ans: List[List[int]]) -> None:
        if not root:
            return
        if sum == root.val and not root.left and not root.right:
            ans.append(curr + [root.val])
            return

        self.dfs(root.left, sum - root.val, curr + [root.val], ans)
        self.dfs(root.right, sum - root.val, curr + [root.val], ans)
